bin/monitor.ml

```diff
@@ -0,0 +1,131 @@
+  match poller with
+  | Error e -> raise (Dune_rpc_private.Version_error.E e)
+  | Ok poller -> poller
+
+(** Shared state *)
+
+let job_number : int Fiber.Svar.t = Fiber.Svar.create 0
+
+let all_diagnostics : Dune_rpc.Diagnostic.Set.t Fiber.Svar.t =
+  Fiber.Svar.create Dune_rpc.Diagnostic.Set.empty
+
+(** Functions on state *)
+
+let display_diagnostics () =
+  Console.reset ();
+  (* TODO should this be parallel or sequential? *)
+  Fiber.parallel_iter_set
+    (module Dune_rpc.Diagnostic.Set)
+    (Fiber.Svar.read all_diagnostics)
+    ~f:(fun err ->
+      Fiber.return
+      @@ Console.print_user_message (Dune_rpc.Diagnostic.to_user_message err))
+
+let maybe_reset_diagnostics progress =
+  match (progress : Dune_rpc.Progress.t) with
+  | Interrupted ->
+    Console.reset ();
+    Fiber.Svar.write all_diagnostics Dune_rpc.Diagnostic.Set.empty
+  | _ -> Fiber.return ()
+
+(** Loops on poll updates *)
+
+let rec progress_loop (progress_poller : Dune_rpc.Progress.t Client.Stream.t) =
+  let* progress = next_exn progress_poller in
+  let job_number = Fiber.Svar.read job_number in
+  status progress job_number;
+  maybe_reset_diagnostics progress >>> progress_loop progress_poller
+
+let rec job_loop (jobs_poller : Dune_rpc.Job.Event.t list Client.Stream.t) =
+  let* jobs = next_exn jobs_poller in
+  Fiber.Svar.write job_number (List.length jobs) >>> job_loop jobs_poller
+
+let rec diagnostic_loop
+    (diagnostic_poller : Dune_rpc.Diagnostic.Event.t list Client.Stream.t) =
+  let* diagnostics = next_exn diagnostic_poller in
+  Fiber.parallel_iter diagnostics ~f:(fun event ->
+      match (event : Dune_rpc.Diagnostic.Event.t) with
+      (* Add diagnostic *)
+      | Add err ->
+        let+ () =
+          Fiber.Svar.write all_diagnostics
+            (Dune_rpc.Diagnostic.Set.add (Fiber.Svar.read all_diagnostics) err)
+        in
+        Console.print_user_message (Dune_rpc.Diagnostic.to_user_message err)
+      (* Remove diagnostic *)
+      | Remove err ->
+        Fiber.Svar.write all_diagnostics
+          (Dune_rpc.Diagnostic.Set.remove (Fiber.Svar.read all_diagnostics) err)
+        >>= display_diagnostics)
+  >>> diagnostic_loop diagnostic_poller
+
+(** Main method *)
+
+let exec () =
+  let where = Rpc_common.active_server () in
+  let* connect = Client.Connection.connect_exn where in
+  Dune_rpc_impl.Client.client connect
+    (Dune_rpc_private.Initialize.Request.create
+       ~id:(Dune_rpc_private.Id.make (Sexp.Atom "monitor_cmd")))
+    ~f:(fun client ->
+      let* progress_poller =
+        poll_exn client Dune_rpc_private.Public.Sub.progress
+      and* jobs_poller =
+        poll_exn client Dune_rpc_private.Public.Sub.running_jobs
+      and* diagnostics_poller =
+        poll_exn client Dune_rpc_private.Public.Sub.diagnostic
+      in
+      Fiber.all_concurrently_unit
+        [ job_loop jobs_poller
+        ; progress_loop progress_poller
+        ; diagnostic_loop diagnostics_poller
+        ])
```

Comment:
This means that you will block until all 3 will have a pending message. You need to have them write to an mvar all concurrently.

Reply:
How so? Are they all not looping independently?

Reply:
Ah yes, I misread this `foo_loop` function.